File: app/src/main/java/com/album/sample/MainActivity.java
package com.album.sample;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Toast;

import com.album.Album;
import com.album.customize.AlbumCameraListener;
import com.album.AlbumConfig;
import com.album.AlbumConstant;
import com.album.AlbumListener;
import com.album.model.AlbumModel;
import com.album.sample.camera.SimpleCameraActivity;
import com.album.ui.annotation.PermissionsType;
import com.album.ui.widget.OnEmptyClickListener;
import com.album.util.AlbumTool;
import com.album.util.FileUtils;
import com.album.util.PermissionUtils;
import com.album.util.scanner.SingleMediaScanner;
import com.album.util.scanner.SingleScannerListener;
import com.yalantis.ucrop.UCrop;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements OnClickListener {

    private UCrop.Options dayOptions;
    private UCrop.Options nightOptions;
    private ArrayList<AlbumModel> list;
    private Uri imagePath;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.btn_day_album).setOnClickListener(this);
        findViewById(R.id.btn_night_album).setOnClickListener(this);
        findViewById(R.id.btn_open_camera).setOnClickListener(this);
        findViewById(R.id.btn_sample_ui).setOnClickListener(this);
        findViewById(R.id.btn_customize_camera).setOnClickListener(this);
        findViewById(R.id.btn_video).setOnClickListener(this);

        dayOptions = new UCrop.Options();
        dayOptions.setToolbarTitle("DayTheme");
        dayOptions.setToolbarColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarBackgroundDay));
        dayOptions.setStatusBarColor(ContextCompat.getColor(this, R.color.colorAlbumStatusBarColorDay));
        dayOptions.setActiveWidgetColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarBackgroundDay));

        nightOptions = new UCrop.Options();
        nightOptions.setToolbarTitle("NightTheme");
        nightOptions.setToolbarWidgetColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarTextColorNight));
        nightOptions.setToolbarColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarBackgroundNight));
        nightOptions.setActiveWidgetColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarBackgroundNight));
        nightOptions.setStatusBarColor(ContextCompat.getColor(this, R.color.colorAlbumStatusBarColorNight));
        list = new ArrayList<>();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_day_album:
                Album
                        .getInstance()
//                        .setAlbumImageLoader(new SimpleFrescoAlbumImageLoader())
                        .setAlbumListener(new MainAlbumListener(this, list))
                        .setAlbumModels(list)
                        .setOptions(dayOptions)
                        .setEmptyClickListener(new OnEmptyClickListener() {
                            @Override
                            public boolean click(View view) {
                                return true;
                            }
                        })
                        .setAlbumCameraListener(null)
                        .setAlbumClass(null)
                        .setConfig(new AlbumConfig()
                                .setCameraCrop(false)
                                .setFilterImg(true)
                                .setPermissionsDeniedFinish(false)
                                .setPreviewFinishRefresh(true)
                                .setAlbumBottomFinderTextBackground(R.drawable.selector_btn)
//                                .setAlbumBottomPreviewTextBackground(R.drawable.selector_btn)
                                .setAlbumBottomSelectTextBackground(R.drawable.selector_btn)
                                .setAlbumContentItemCheckBoxDrawable(R.drawable.simple_selector_album_item_check)
//                                .setFrescoImageLoader(true)  // 通知 Album 图片加载框架使用的是 Fresco
                                .setPreviewBackRefresh(true))
                        .start(this);
                break;
            case R.id.btn_night_album:
                Album
                        .getInstance()
                        .setAlbumListener(new MainAlbumListener(this, null))
                        .setOptions(nightOptions)
                        .setAlbumCameraListener(null)
                        .setAlbumClass(null)
                        .setEmptyClickListener(new OnEmptyClickListener() {
                            @Override
                            public boolean click(View view) {
                                return true;
                            }
                        })
                        .setConfig(new AlbumConfig(AlbumConstant.TYPE_NIGHT)
                                .setRadio(true)
                                .setFilterImg(false)
                                .setHideCamera(true)
                                .setCrop(true).setCameraPath(Environment.getExternalStorageDirectory().getPath() + "/" + "DCIM/Album")
                                .setUCropPath(Environment.getExternalStorageDirectory().getPath() + "/" + "DCIM" + "/" + "uCrop")
                                .setPermissionsDeniedFinish(true)
                                .setPreviewFinishRefresh(true)
                                .setPreviewBackRefresh(true)
                                .setCameraCrop(true))
                        .start(this);
                break;

            case R.id.btn_sample_ui:

                Album
                        .getInstance()
                        .setAlbumListener(new MainAlbumListener(this, list))
                        .setAlbumClass(SimpleAlbumUI.class)
                        .setAlbumCameraListener(null)
                        .setEmptyClickListener(null)
                        .setConfig(new AlbumConfig()
                                .setCameraCrop(false)
                                .setAlbumPreviewBackground(R.color.colorAlbumPreviewBackgroundNight)
                                .setAlbumToolbarBackground(R.color.colorPrimary)
                                .setAlbumStatusBarColor(R.color.colorPrimary)
                                .setPreviewBackRefresh(true))
                        .start(this);

                break;

            case R.id.btn_open_camera:
                // fragment  activity 直接传递 this ， 内部会自己处理
                int i = AlbumTool.openCamera(this, imagePath = Uri.fromFile(FileUtils.getCameraFile(this, null, false)), false);

                // -1 没有权限
                // 0 打开成功
                // 1 打开相机失败，估计设备没有安装相关软件

                break;

            case R.id.btn_customize_camera:

                Album
                        .getInstance()
                        .setAlbumListener(new MainAlbumListener(this, list))
                        .setAlbumModels(list)
                        .setOptions(dayOptions)
                        .setAlbumCameraListener(new AlbumCameraListener() {
                            @Override
                            public void startCamera(@NonNull Fragment fragment) {
                                if (PermissionUtils.storage(fragment.getActivity()) && PermissionUtils.camera(fragment.getActivity())) {
                                    FragmentActivity activity = fragment.getActivity();
                                    Toast.makeText(activity, "camera", Toast.LENGTH_SHORT).show();
                                    Intent intent = new Intent(activity, SimpleCameraActivity.class);
                                    fragment.startActivityForResult(intent, AlbumConstant.CUSTOMIZE_CAMERA_RESULT_CODE);
                                }
                            }
                        })
                        .setEmptyClickListener(new OnEmptyClickListener() {
                            @Override
                            public boolean click(View view) {
                                return true;
                            }
                        })
                        .setAlbumClass(null)
                        .setConfig(new AlbumConfig()
                                .setCameraCrop(false)
                                .setPermissionsDeniedFinish(false)
                                .setPreviewFinishRefresh(true)
                                .setAlbumBottomFinderTextBackground(R.drawable.selector_btn)
//                                .setAlbumBottomPreviewTextBackground(R.drawable.selector_btn)
                                .setAlbumBottomSelectTextBackground(R.drawable.selector_btn)
                                .setAlbumContentItemCheckBoxDrawable(R.drawable.simple_selector_album_item_check)
//                                .setFrescoImageLoader(true)  // 通知 Album 图片加载框架使用的是 Fresco
                                .setPreviewBackRefresh(true))
                        .start(this);

                break;

            case R.id.btn_video:
                Album
                        .getInstance()
                        .setAlbumListener(new MainAlbumListener(this, list))
                        .setAlbumCameraListener(null)
                        .setEmptyClickListener(new OnEmptyClickListener() {
                            @Override
                            public boolean click(View view) {
                                return true;
                            }
                        })
                        .setConfig(
                                new AlbumConfig()
                                        .setVideo(true)
                                        .setAlbumToolbarText(R.string.album_video_title)
                                        .setAlbumContentViewCameraTips(R.string.video_tips)
                                        .setPreviewBackRefresh(true))
                        .start(this);
                break;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_CANCELED) {
        } else if (resultCode == UCrop.RESULT_ERROR) {
        } else if (resultCode == Activity.RESULT_OK) {
            switch (requestCode) {
                case AlbumConstant.ITEM_CAMERA:
                    new SingleMediaScanner(this, FileUtils.getScannerFile(imagePath.getPath()),
                            new SingleScannerListener() {
                                @Override
                                public void onScanStart() {

                                }

                                @Override
                                public void onScanCompleted(int type) {

                                }
                            }, AlbumConstant.TYPE_RESULT_CAMERA);
                    UCrop.of(Uri.fromFile(new File(imagePath.getPath())), imagePath = Uri.fromFile(FileUtils.getCameraFile(this, null, false)))
                            .withOptions(new UCrop.Options())
                            .start(this);
                    break;
                case UCrop.REQUEST_CROP:
                    new SingleMediaScanner(this, FileUtils.getScannerFile(imagePath.getPath()),
                            new SingleScannerListener() {
                                @Override
                                public void onScanStart() {

                                }

                                @Override
                                public void onScanCompleted(int type) {

                                }
                            }, AlbumConstant.TYPE_RESULT_CROP);
                    Toast.makeText(getApplicationContext(), imagePath.getPath(), Toast.LENGTH_SHORT).show();
                    break;
            }
        }
    }

    /**
     * @see com.album.ui.widget.SimpleAlbumListener
     */
    private static class MainAlbumListener implements AlbumListener {

        private Context context;
        private List<AlbumModel> list = null;

        void toast(String s) {
            Toast.makeText(context, s, Toast.LENGTH_SHORT).show();
        }

        MainAlbumListener(Context context, ArrayList<AlbumModel> list) {
            this.context = context.getApplicationContext();
            this.list = list;
        }

        @Override
        public void onAlbumActivityFinish() {
            toast("album activity finish");
        }

        @Override
        public void onAlbumPermissionsDenied(@PermissionsType int type) {
            toast("permissions error");
        }

        @Override
        public void onAlbumFragmentNull() {
            toast("album fragment null");
        }

        @Override
        public void onAlbumPreviewFileNull() {
            toast("preview image has been deleted");
        }

        @Override
        public void onAlbumFinderNull() {
            toast("folder directory is empty");
        }

        @Override
        public void onAlbumBottomPreviewNull() {
            toast("preview no image");
        }

        @Override
        public void onAlbumBottomSelectNull() {
            toast("select no image");
        }

        @Override
        public void onAlbumFragmentFileNull() {
            toast("album image has been deleted");
        }

        @Override
        public void onAlbumPreviewSelectNull() {
            toast("PreviewActivity,  preview no image");
        }

        @Override
        public void onAlbumCheckBoxFileNull() {
            toast("check box  image has been deleted");
        }

        @Override
        public void onAlbumFragmentCropCanceled() {
            toast("cancel crop");
        }

        @Override
        public void onAlbumFragmentCameraCanceled() {
            toast("cancel camera");
        }

        @Override
        public void onAlbumFragmentUCropError(@Nullable Throwable data) {
            Log.i("ALBUM", data.getMessage());
            toast("crop error:" + data);
        }

        @Override
        public void onAlbumResources(@NonNull List<AlbumModel> list) {
            toast("select count :" + list.size());
            if (this.list != null) {
                this.list.clear();
                this.list.addAll(list);
            }
        }

        @Override
        public void onAlbumUCropResources(@Nullable File scannerFile) {
            toast("crop file:" + scannerFile);
        }

        @Override
        public void onAlbumMaxCount() {
            toast("select max count");
        }

        @Override
        public void onAlbumActivityBackPressed() {
            toast("AlbumActivity Back");
        }

        @Override
        public void onAlbumOpenCameraError() {
            toast("camera error");
        }

        @Override
        public void onAlbumEmpty() {
            toast("no image");
        }

        @Override
        public void onAlbumNoMore() {
            toast("album no more");
        }

        @Override
        public void onAlbumResultCameraError() {
            toast("result error");
        }

        @Override
        public void onVideoPlayError() {
            toast("play video error : checked video app");
        }
    }

}

File: app/src/main/java/com/album/sample/MainActivity.kt
@file:Suppress("FunctionName")

package com.album.sample

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.View
import android.view.View.OnClickListener
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.album.core.Album
import com.album.core.AlbumBundle
import com.album.core.AlbumCameraConst
import com.album.core.AlbumConst
import com.album.core.action.AlbumImageLoader
import com.album.core.ext.albumPathFile
import com.album.core.ext.openCamera
import com.album.core.ext.permission.permissionCamera
import com.album.core.ext.permission.permissionStorage
import com.album.core.ext.uri
import com.album.sample.camera.SimpleCameraActivity
import com.album.sample.imageloader.SimpleGlideImageLoader
import com.album.sample.imageloader.SimplePicassoAlbumImageLoader
import com.album.sample.imageloader.SimpleSubsamplingScaleImageLoader
import com.gallery.scan.ScanEntity
import com.gallery.scan.SingleMediaScanner
import com.gallery.scan.args.ScanConst
import com.album.ui.AlbumUiBundle
import com.album.ui.ui
import com.yalantis.ucrop.UCrop
import com.yalantis.ucrop.UCropFragment
import com.yalantis.ucrop.UCropFragmentCallback
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

fun NightAlbumBundle(): AlbumBundle {
    return AlbumBundle(
            spanCount = 4,
            cropFinish = false,
            checkBoxDrawable = R.drawable.simple_selector_album_item_check,
            radio = true,
            cameraPath = Environment.getExternalStorageDirectory().path + "/" + "DCIM/Album",
            uCropPath = Environment.getExternalStorageDirectory().path + "/" + "DCIM" + "/" + "uCrop",
            cameraTextColor = R.color.colorAlbumContentViewTipsColorNight,
            cameraDrawable = R.drawable.ic_camera_drawable,
            cameraDrawableColor = R.color.colorAlbumContentViewCameraDrawableColorNight,
            cameraBackgroundColor = R.color.colorAlbumToolbarBackgroundNight,
            rootViewBackground = R.color.colorAlbumContentViewBackgroundNight,
            cameraCrop = true)
}

fun NightAlbumUIBundle(): AlbumUiBundle {
    return AlbumUiBundle(
            statusBarColor = R.color.colorAlbumStatusBarColorNight,
            toolbarBackground = R.color.colorAlbumToolbarBackgroundNight,
            toolbarIconColor = R.color.colorAlbumToolbarIconColorNight,
            toolbarTextColor = R.color.colorAlbumToolbarTextColorNight,
            bottomFinderTextBackground = R.color.colorAlbumBottomViewBackgroundNight,
            bottomFinderTextColor = R.color.colorAlbumBottomFinderTextColorNight,
            bottomFinderTextDrawableColor = R.color.colorAlbumBottomFinderTextDrawableColorNight,
            bottomPreViewTextColor = R.color.colorAlbumBottomPreViewTextColorNight,
            bottomSelectTextColor = R.color.colorAlbumBottomSelectTextColorNight,
            listPopupBackground = R.color.colorAlbumListPopupBackgroundNight,
            listPopupItemTextColor = R.color.colorAlbumListPopupItemTextColorNight,
            preBackground = R.color.colorAlbumPreviewBackgroundNight,
            preBottomViewBackground = R.color.colorAlbumPreviewBottomViewBackgroundNight,
            preBottomOkTextColor = R.color.colorAlbumPreviewBottomViewOkColorNight,
            preBottomCountTextColor = R.color.colorAlbumPreviewBottomViewCountColorNight)
}

fun MainActivity.dayAlbum() {
    Album.instance.apply {
        albumImageLoader = SimplePicassoAlbumImageLoader()
        albumListener = MainAlbumListener(applicationContext, list)
        selectList = list
        options = dayOptions
    }.ui(this,
            AlbumBundle(scanType = ScanConst.IMAGE, checkBoxDrawable = R.drawable.simple_selector_album_item_check))
}

fun MainActivity.nightAlbum() {
    Album.instance.apply {
        albumListener = MainAlbumListener(applicationContext, null)
        options = nightOptions
        albumImageLoader = SimplePicassoAlbumImageLoader()
        customCameraListener = {
            if (it.permissionStorage() && it.permissionCamera()) {
                Toast.makeText(it.activity, "camera", Toast.LENGTH_SHORT).show()
                val intent = Intent(it.activity, SimpleCameraActivity::class.java)
                it.startActivityForResult(intent, AlbumCameraConst.CUSTOM_CAMERA_REQUEST_CODE)
            }
        }
    }.ui(this, NightAlbumBundle(), NightAlbumUIBundle())
}

fun MainActivity.video() {
    Album.instance.apply {
        albumImageLoader = SimplePicassoAlbumImageLoader()
        albumListener = MainAlbumListener(applicationContext, null)
    }.ui(this, AlbumBundle(
            scanType = ScanConst.VIDEO,
            cameraText = R.string.video_tips),
            AlbumUiBundle(toolbarText = R.string.album_video_title))
}

fun MainActivity.startCamera() {
    imagePath = uri(applicationContext.albumPathFile(null, System.currentTimeMillis().toString(), "jpg"))
    val i = openCamera(imagePath, false)
    Log.d(javaClass.simpleName, i.toString())
}

fun MainActivity.imageLoader(imageLoader: AlbumImageLoader) {
    Album.instance.apply { albumImageLoader = imageLoader }.ui(this)
}

class SimpleSingleScannerListener : SingleMediaScanner.SingleScannerListener {
    override fun onScanCompleted(type: Int, path: String) {}
    override fun onScanStart() {}
}

class MainActivity : AppCompatActivity(), OnClickListener, UCropFragmentCallback {

    lateinit var dayOptions: UCrop.Options
    lateinit var nightOptions: UCrop.Options
    lateinit var list: ArrayList<ScanEntity>
    lateinit var imagePath: Uri

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btn_day_album.setOnClickListener(this)
        btn_night_album.setOnClickListener(this)
        btn_open_camera.setOnClickListener(this)
        btn_video.setOnClickListener(this)
        btn_imageloader.setOnClickListener(this)

        dayOptions = UCrop.Options()
        dayOptions.apply {
            setToolbarTitle("DayTheme")
            setToolbarColor(ContextCompat.getColor(this@MainActivity, R.color.colorAlbumToolbarBackground))
            setStatusBarColor(ContextCompat.getColor(this@MainActivity, R.color.colorAlbumStatusBarColor))
            setActiveWidgetColor(ContextCompat.getColor(this@MainActivity, R.color.colorAlbumToolbarBackground))
        }

        nightOptions = UCrop.Options()
        nightOptions.setToolbarTitle("NightTheme")
        nightOptions.setToolbarWidgetColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarTextColorNight))
        nightOptions.setToolbarColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarBackgroundNight))
        nightOptions.setActiveWidgetColor(ContextCompat.getColor(this, R.color.colorAlbumToolbarBackgroundNight))
        nightOptions.setStatusBarColor(ContextCompat.getColor(this, R.color.colorAlbumStatusBarColorNight))

        list = ArrayList()
    }

    override fun onClick(v: View) {
        Album.destroy()
        when (v.id) {
            R.id.btn_day_album -> dayAlbum()
            R.id.btn_night_album -> nightAlbum()
            R.id.btn_open_camera -> startCamera()
            R.id.btn_video -> video()
            R.id.btn_imageloader -> AlertDialog.Builder(this@MainActivity)
                    .setSingleChoiceItems(arrayOf("Glide", "Picasso", "SubsamplingScale"), -1
                    ) { dialog, which ->
                        when (which) {
                            0 -> imageLoader(SimpleGlideImageLoader())
                            1 -> imageLoader(SimplePicassoAlbumImageLoader())
                            2 -> imageLoader(SimpleSubsamplingScaleImageLoader())
                        }
                        dialog.dismiss()
                    }.show()
        }
    }

    override fun onCropFinish(result: UCropFragment.UCropResult) {
    }

    override fun loadingProgress(showLoader: Boolean) {
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (resultCode) {
            Activity.RESULT_CANCELED -> {
            }
            UCrop.RESULT_ERROR -> {
            }
            Activity.RESULT_OK -> when (requestCode) {
                AlbumCameraConst.CAMERA_REQUEST_CODE -> {
                    SingleMediaScanner(this, imagePath.path
                            ?: "", AlbumConst.TYPE_RESULT_CAMERA, SimpleSingleScannerListener())
                }
                UCrop.REQUEST_CROP -> {
                    SingleMediaScanner(this, imagePath.path
                            ?: "", AlbumConst.TYPE_RESULT_CROP, SimpleSingleScannerListener())
                    Toast.makeText(applicationContext, imagePath.path, Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

}
